### PR TITLE
docker: support url in addition to registry-url/image

### PR DESCRIPTION
**Pull request by abderrahim**
Mirroring is currently almost impossible as it requires the mirrored image
to have the exact same image name as the original. That is, I cannot mirror
registry.hub.docker.com/buildstream/buildstream to
myregistry.com/myproject/buildstream/buildstream

### CODE REVIEW

**Review by nanonyme — APPROVED** [review]
Looks good to me. I'm a bit worried if it's legal to have registry which is mounted deeper in URL (IOW first path segment is not v2)

**Comment by abderrahim** [comment]
> Looks good to me. I'm a bit worried if it's legal to have registry which is mounted deeper in URL (IOW first path segment is not v2)

Yeah, I don't see that when reading the [spec](https://distribution.github.io/distribution/spec/api/). But [the issue](https://gitlab.com/BuildStream/bst-external/-/issues/9) that led to this implementation seems to imply otherwise.

**Comment by wsalmonct** [comment]
Does this silently ignore `image` and `registry-url` if `url` is present?

It raises a error if nether are present. but should it not also raise a error if both are?

The condition of both `image` and `url` does not cause a error as such but it does cause unexpected behaviour. usually if you provide a key that is not used then bst causes a error (node.validate_keys) so i think the extra error would follow the existing patten?

**Comment by abderrahim** [comment]
> Does this silently ignore `image` and `registry-url` if `url` is present?
> 
> It raises a error if nether are present. but should it not also raise a error if both are?

Seems you're right. Somehow I thought the current condition would handle it.

> The condition of both `image` and `url` does not cause a error as such but it does cause unexpected behaviour. usually if you provide a key that is not used then bst causes a error (node.validate_keys) so i think the extra error would follow the existing patten?

I agree.

**Comment by wsalmonct** [comment]
The new version looks good to me!

I don't have permission to approve but i would if i could.

File: src/buildstream_plugins/sources/docker.py [hunk 1]
```diff
@@ -37,6 +37,10 @@
    # Image path (required)
```
Comment (gtristan) [inline src/buildstream_plugins/sources/docker.py:37]: Nit: We should at least remove the "(required)" part of the docs here.

**Review by gtristan — APPROVED** [review]
Asides from the docs related comment, this looks correct to me.

Note that we will have to bump the buildstream-plugins minor point to `2.2.0` for this API addition.

**Pull request merged**